src/array.jl

```diff
@@ -494,6 +494,11 @@ similar(A::CategoricalArray{S, M, R}, ::Type{T},
         dims::NTuple{N, Int}) where {S, T<:Union{CatValue, Missing}, M, N, R} =
     CategoricalArray{Union{T, Missing}, N, R}(undef, dims)
 
+similar(::Type{T}, dims::Dims) where {T<:AbstractArray{<:Union{CatValue,Missing}}} =
+    CategoricalArray{eltype(T)}(undef, dims)
+similar(::Type{T}, dims::Dims) where {T<:AbstractArray{Missing}} =
```

Comment:
Do you need it to avoid method ambiguity? If yes - maybe it would be better to define four methods with categorical values explicitly and avoid type piracy?

Reply:
The problem is that `<:Union{CatValue,Missing}` includes `Missing`, which is completely different. Annoyingly, the only way to avoid the type piracy is to define methods separately for `CategoricalString` and `CategoricalValue`. But that's probably better, and anyway we should deprecate `CategoricalString`.